src/ray/raylet/scheduling/policy/node_label_scheduling_policy.cc

```diff
@@ -0,0 +1,195 @@
+
+namespace ray {
+namespace raylet_scheduling_policy {
+
+scheduling::NodeID NodeLabelSchedulingPolicy::Schedule(
+    const ResourceRequest &resource_request, SchedulingOptions options) {
+  RAY_CHECK(options.scheduling_type == SchedulingType::NODE_LABEL);
+  auto context =
+      dynamic_cast<const NodeLabelSchedulingContext *>(options.scheduling_context.get());
+  const auto &scheduling_strategy = context->GetSchedulingStrategy();
+  RAY_CHECK(scheduling_strategy.has_node_label_scheduling_strategy());
+  const auto &node_label_scheduling_strategy =
+      scheduling_strategy.node_label_scheduling_strategy();
+  // 1. Selet feasible nodes
+  auto candidate_nodes = SelectFeasibleNodes(resource_request);
+  if (candidate_nodes.empty()) {
+    return scheduling::NodeID::Nil();
+  }
+  // 2. Filter by hard expressions.
+  if (node_label_scheduling_strategy.hard().expressions().size() > 0) {
+    FilterNodesByHardExpressions(candidate_nodes, node_label_scheduling_strategy.hard());
+    if (candidate_nodes.empty()) {
+      return scheduling::NodeID::Nil();
+    }
+  }
+
+  // 3. Filter by soft expressions.
+  const auto &soft_expressions = node_label_scheduling_strategy.soft();
+  if (soft_expressions.expressions().size() > 0) {
+    auto match_all_soft_nodes =
+        FilterNodesBySoftExpressions(candidate_nodes, soft_expressions);
+    if (!match_all_soft_nodes.empty()) {
+      return SelectBestNode(match_all_soft_nodes, resource_request);
+    }
+  }
+
+  return SelectBestNode(candidate_nodes, resource_request);
+}
+
+scheduling::NodeID NodeLabelSchedulingPolicy::SelectBestNode(
+    const absl::flat_hash_map<scheduling::NodeID, const Node *> &candidate_nodes,
+    const ResourceRequest &resource_request) {
+  auto available_nodes = SelectAvailableNodes(candidate_nodes, resource_request);
+  if (available_nodes.empty()) {
+    return SelectRandomNode(candidate_nodes);
+  }
+  return SelectRandomNode(available_nodes);
+}
+
+scheduling::NodeID NodeLabelSchedulingPolicy::SelectRandomNode(
+    const absl::flat_hash_map<scheduling::NodeID, const Node *> &candidate_nodes) {
+  std::uniform_int_distribution<int> distribution(0, candidate_nodes.size() - 1);
+  int node_index = distribution(gen_);
+  auto it = candidate_nodes.begin();
+  std::advance(it, node_index);
+  return it->first;
+}
+
+void NodeLabelSchedulingPolicy::FilterNodesByHardExpressions(
+    absl::flat_hash_map<scheduling::NodeID, const Node *> &candidate_nodes,
+    const rpc::LabelMatchExpressions &expressions) const {
+  for (const auto &expression : expressions.expressions()) {
+    if (candidate_nodes.empty()) {
+      return;
+    }
+
+    for (auto iter = candidate_nodes.begin(); iter != candidate_nodes.end(); iter++) {
+      if (!IsNodeMatchLabelExpression(*(iter->second), expression)) {
+        candidate_nodes.erase(iter);
+      }
+    }
+  }
+}
+
+absl::flat_hash_map<scheduling::NodeID, const Node *>
+NodeLabelSchedulingPolicy::FilterNodesBySoftExpressions(
```

Comment:
I think we can combine `FilterNodesByHardExpressions` and `FilterNodesBySoftExpressions` into a single function  `FilterNodesByLabelMatchExpressions` that takes nodes and returns filtered nodes. They basically have the same implementation.

Reply:
Fixed.